packages/flutter_tools/lib/src/reporting/events.dart

```diff
@@ -249,3 +249,40 @@ class CodeSizeEvent extends UsageEvent {
 class ErrorHandlingEvent extends UsageEvent {
   ErrorHandlingEvent(String parameter) : super('error-handling', parameter, flutterUsage: globals.flutterUsage);
 }
+
+/// The category for analytics events related to null safety.
+const String kNullSafetyCategory = 'null-safety';
+
+/// Emit various null safety analytic events.
+///
+/// 1. The current null safety runtime mode.
+/// 2. The % of dependencies that are not migrated.
+/// 3. The main language version.
+void collectLanguageVersionEvents(
+  PackageConfig packageConfig,
+  NullSafetyMode nullSafetyMode,
+  String currentPackage,
+  Usage flutterUsage,
+) {
+  if (packageConfig.packages.isEmpty) {
+    return;
+  }
+  int migrated = 0;
+  LanguageVersion languageVersion;
+  for (final Package package in packageConfig.packages) {
+    if (package.name == currentPackage) {
+      languageVersion = package.languageVersion;
+    }
+    if (package.languageVersion.major >= nullSafeVersion.major &&
+        package.languageVersion.minor >= nullSafeVersion.minor) {
+      migrated += 1;
+    }
+  }
+  final int formattedPercentage = (migrated / packageConfig.packages.length * 100).round();
+  flutterUsage.sendEvent(kNullSafetyCategory, 'runtime-mode', label: nullSafetyMode.toString());
+  flutterUsage.sendEvent(kNullSafetyCategory, 'migrated', value: formattedPercentage);
```

Comment:
The analytics web interface will end up summing the reported values, which would not be useful here, but it will also give an average, which makes more sense. If you report it as a custom dimension, the analytics web interface won't do anything very useful with it, but since the data is there, you can write database queries against it. Another option would be to bucket the values (10, 20, 30%), and report the bucketed values in the custom dimension.

An orthogonal question would be whether you want to weight the percentage by the number of libraries. For that, it might be easier to report both the number of migrated packages and the total number of packages, and then compute aggregates with a database query after the fact.

Reply:
Seems like that would be easier to do if both migrated and total libraries were custom dimensions on a different event?

Reply:
Yeah, like:
```dart
flutterUsage.sendEvent(kNullSafetyCategory, 'stats', parameters: {
  migratedLibraries: migrated,
  totalLibraries: total,
});
```